extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -234,10 +253,55 @@ private Map<URI, FileSizeInfo> getFileSizeCache() {
         return cache;
     }
 
+    /**
+     * Fetch the size of the file at the given S3 URI.
+     *
+     * @throws NoSuchKeyException if the file does not exist
+     * @throws IOException if there is an error fetching the file size
+     */
+    long fetchFileSize(@NotNull final S3Uri s3Uri) throws IOException {
+        final long cachedSize = getCachedSize(s3Uri.uri());
+        if (cachedSize >= 0) {
+            return cachedSize;
+        }
+        // Fetch the size of the file using a blocking HEAD request, and store it in the cache for future use
+        if (log.isDebugEnabled()) {
+            log.debug().append("Head: ").endl();
+        }
+        final HeadObjectResponse headObjectResponse;
+        try {
+            headObjectResponse = s3AsyncClient
+                    .headObject(HeadObjectRequest.builder()
+                            .bucket(s3Uri.bucket().orElseThrow())
+                            .key(s3Uri.key().orElseThrow())
+                            .build())
+                    .get(s3Instructions.readTimeout().toNanos(), TimeUnit.NANOSECONDS);
+        } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+            throw handleS3Exception(e, String.format("fetching HEAD for file %s", s3Uri), s3Instructions);
+        }
+        final long fileSize = headObjectResponse.contentLength();
+        updateFileSizeCache(s3Uri.uri(), fileSize);
+        return fileSize;
+    }
+
+    /**
+     * Get the cached size for the given URI, or -1 if the size is not cached.
+     */
+    private long getCachedSize(final URI uri) {
+        final Map<URI, FileSizeInfo> fileSizeCache = fileSizeCacheRef.get();
+        if (fileSizeCache != null) {
+            final FileSizeInfo sizeInfo = fileSizeCache.get(uri);
+            if (sizeInfo != null) {
+                return sizeInfo.size;
+            }
+        }
+        return -1;
```

Comment:
I would define a constant for this. `NULL_SIZE ` or `UNKNOWN_SIZE`. Then compare `==` to the constant, instead of `<0` or `>=0`.